File: src/lib/theme.ts
/**
 * Which palette the window paints in, and whose colours those are.
 *
 * Two questions, kept apart. `Theme` is light, dark, or a standing instruction
 * to keep asking the OS — which is why resolving it lives here in JS rather
 * than in a media query. The stylesheet only ever sees the answer, since
 * `data-theme` on <html> is always "light" or "dark", so the light palette is
 * written once instead of once per route to it.
 *
 * A [`CustomTheme`] is the second question: whose ink, whose accent, whose
 * typeface and wordmark. It is applied *over* whichever of the two palettes is
 * showing, as inline custom properties on <html> — which beat both the `:root`
 * block and the `[data-theme="light"]` one on specificity without either of
 * them having to know custom themes exist. The whole of what makes that
 * possible is a stylesheet that names its raw values exactly once and speaks in
 * roles everywhere else: fourteen colours here move six thousand lines below.
 *
 * Both settings belong to Rust, in the same `settings.json` as everything else
 * the app remembers. What is kept here is a copy, and only to cover the gap:
 * settings arrive over IPC a tick or two after the window paints, and somebody
 * who chose light on a dark-mode machine — or who branded the app violet —
 * should not watch it open wrong and correct itself every single time.
 */

import type { CustomTheme, Theme } from "./api";

const CACHE_KEY = "taurus.theme";
/** The resolved custom properties from last time, for the frame before IPC. */
const TOKENS_KEY = "taurus.themeTokens";
const DARK_QUERY = "(prefers-color-scheme: dark)";

export type ResolvedTheme = "light" | "dark";

export function systemTheme(): ResolvedTheme {
  return window.matchMedia?.(DARK_QUERY).matches ? "dark" : "light";
}

export function resolve(theme: Theme): ResolvedTheme {
  return theme === "system" ? systemTheme() : theme;
}

/**
 * Which palette is actually painted, once a custom theme has had its say.
 *
 * A theme that fills in only one of the two is a statement — this brand is
 * dark — and honouring the preference anyway would paint half a palette: the
 * light surfaces the theme never gave, under the dark text it did. So the
 * theme wins, and the UI says so rather than leaving a preference that appears
 * to do nothing. See `ThemePicker`.
 */
export function resolveWith(theme: Theme, custom: CustomTheme | null): ResolvedTheme {
  switch (custom?.modes) {
    case "dark_only":
      return "dark";
    case "light_only":
      return "light";
    default:
      return resolve(theme);
  }
}

/** Whether `mode` is one this theme can actually paint. */
export function supports(custom: CustomTheme | null, mode: ResolvedTheme): boolean {
  return custom === null || custom.modes === "both" || custom.modes === `${mode}_only`;
}

/**
 * What a theme file's colour names are called in the stylesheet.
 *
 * The right-hand side is the design system's own vocabulary, which names its
 * accents after the colours they happen to be. That is fine for a system with
 * one palette and absurd in a theme file, where the entire point is that the
 * accent might be violet — so a theme names the *job* and this table is where
 * the two meet. The same table exists in `crates/taurus-host/src/theme.rs`,
 * which is the authority; `theme.test.ts` fails if they drift apart.
 */
export const COLOR_TOKENS: Record<string, string> = {
  ink: "--lk-ink",
  "surface-1": "--lk-surface-1",
  "surface-2": "--lk-surface-2",
  "surface-hover": "--lk-surface-hover",
  line: "--lk-line",
  text: "--lk-text",
  "text-dim": "--lk-text-dim",
  "text-faint": "--lk-text-faint",
  accent: "--lk-cyan",
  "accent-hover": "--lk-cyan-hover",
  "on-accent": "--lk-on-cyan",
  ok: "--lk-mint",
  warn: "--lk-peach",
  danger: "--lk-red",
};

/** The same, for the three typefaces. */
export const FONT_TOKENS: Record<string, string> = {
  display: "--lk-display",
  body: "--lk-body",
  mono: "--lk-mono-face",
};

/**
 * How the colour fields are grouped and ordered in the editor.
 *
 * A flat list of fourteen swatches is a list nobody can hold; four groups of
 * three or four is four decisions. The order within each is the order they
 * stack on screen — the window, then what is raised off it — so picking them
 * top to bottom builds a ladder rather than a set.
 */
export const COLOR_GROUPS: readonly { label: string; hint: string; keys: string[] }[] = [
  {
    label: "Surfaces",
    hint: "The window, the panels raised off it, and the one hairline between them.",
    keys: ["ink", "surface-1", "surface-2", "surface-hover", "line"],
  },
  {
    label: "Text",
    hint: "Three weights. The faint one carries the 10px labels, so it needs the most contrast, not the least.",
    keys: ["text", "text-dim", "text-faint"],
  },
  {
    label: "Accent",
    hint: "The lead colour, its hover, and what stays legible on top of it.",
    keys: ["accent", "accent-hover", "on-accent"],
  },
  {
    label: "Signals",
    hint: "Spent sparingly. A surface that is 5% accent reads as signal; one that is 30% reads as decoration.",
    keys: ["ok", "warn", "danger"],
  },
];

/**
 * The corner-radius ladder as the stylesheet ships it, which `shape.radius`
 * scales. Named here because a multiplier needs something to multiply.
 */
const RADII: Record<string, number> = {
  "--r-sm": 6,
  "--r": 9,
  "--r-md": 10,
  "--r-lg": 12,
  "--r-xl": 14,
};

/** Every custom property a theme is allowed to set. */
function everyToken(): string[] {
  return [
    ...Object.values(COLOR_TOKENS),
    ...Object.values(FONT_TOKENS),
    ...Object.keys(RADII),
    "--gutter",
    "--gutter-rail",
  ];
}

/**
 * The custom properties a theme sets, for one of the two modes.
 *
 * Only what the theme states. Everything it leaves out falls through to the
 * stylesheet, which is what lets the common case — a different accent — be
 * four lines rather than a transcription of the palette, and what keeps a
 * theme written today working after the app adds a token tomorrow.
 */
export function tokensFor(
  custom: CustomTheme | null,
  mode: ResolvedTheme,
): Record<string, string> {
  const out: Record<string, string> = {};
  if (!custom) return out;

  for (const [name, value] of Object.entries(mode === "dark" ? custom.dark : custom.light)) {
    const token = COLOR_TOKENS[name];
    if (token) out[token] = value;
  }

  for (const [name, token] of Object.entries(FONT_TOKENS)) {
    const family = custom.fonts[name as keyof CustomTheme["fonts"]];
    // Quoted, because the stylesheet interpolates this straight into a
    // `font-family` list where a two-word family name is two families.
    if (family) out[token] = /^["']/.test(family) ? family : `"${family}"`;
  }

  const { radius, gutter } = custom.shape;
  if (radius !== null && radius !== undefined) {
    // Rounded to whole pixels: a 9.6px corner and a 9px one beside it are the
    // kind of difference that reads as a mistake rather than as a choice.
    for (const [token, base] of Object.entries(RADII)) {
      out[token] = `${Math.round(base * radius)}px`;
    }
  }
  if (gutter !== null && gutter !== undefined) out["--gutter"] = `${gutter}px`;
  const rail = custom.shape["rail-gutter"];
  if (rail !== null && rail !== undefined) out["--gutter-rail"] = `${rail}px`;

  return out;
}

/**
 * Writes a token set onto <html>, and clears whatever the last one set.
 *
 * Every token is removed before the new set is applied rather than diffed
 * against it. The set is twenty-two properties, so the diff would buy nothing
 * measurable — and the bug it avoids is the one that matters: switching from a
 * theme that sets `--gutter` to one that does not has to give the stylesheet's
 * gutter back, and a version of this that only ever wrote would leave the old
 * one standing with nothing on screen to say where it came from.
 */
function paint(tokens: Record<string, string>): void {
  const style = document.documentElement.style;
  for (const token of everyToken()) style.removeProperty(token);
  for (const [token, value] of Object.entries(tokens)) style.setProperty(token, value);
}

/**
 * Paints `theme` and remembers it for the next cold start.
 *
 * The preference is cached, not the resolution: caching "light" for someone who
 * asked to follow a system that happened to be light at the time would freeze
 * them there through every sunset after. The custom theme's *tokens* are
 * cached, though, and that asymmetry is deliberate — they are the resolution
 * already, they cannot go stale by the clock, and re-deriving them needs a
 * theme file that has not arrived over IPC yet at the moment they are wanted.
 *
 * The logo is not cached. It is the one part that is large, and a mark that
 * appears a tick into the launch is a far smaller thing than a window that
 * opens in the wrong palette.
 */
export function applyTheme(theme: Theme, custom: CustomTheme | null = null): void {
  const mode = resolveWith(theme, custom);
  document.documentElement.dataset.theme = mode;
  const tokens = tokensFor(custom, mode);
  paint(tokens);
  try {
    localStorage.setItem(CACHE_KEY, theme);
    localStorage.setItem(TOKENS_KEY, JSON.stringify(tokens));
  } catch {
    // A webview with storage disabled costs a flash on the next start, which
    // is not worth failing a theme change over.
  }
}

/**
 * Repaints the frame before settings arrive, from what was showing last time.
 *
 * Called once at startup. Without it, a branded app opens in the shipped cyan
 * and corrects itself a round trip later — which is the same flash the theme
 * preference has always been cached to avoid, and more visible, because a
 * whole palette moves rather than one of two.
 */
export function applyCachedTheme(): void {
  const theme = cachedTheme();
  document.documentElement.dataset.theme = resolve(theme);
  try {
    const stored: unknown = JSON.parse(localStorage.getItem(TOKENS_KEY) ?? "{}");
    if (stored && typeof stored === "object" && !Array.isArray(stored)) {
      // Filtered against the allowed set rather than written through. This is
      // the one path where a value reaches a style attribute without having
      // come from Rust's validation this run, and local storage is editable.
      const allowed = new Set(everyToken());
      const tokens: Record<string, string> = {};
      for (const [token, value] of Object.entries(stored as Record<string, unknown>)) {
        if (allowed.has(token) && typeof value === "string") tokens[token] = value;
      }
      paint(tokens);
    }
  } catch {
    // Unreadable or not JSON. The stylesheet's own palette is the answer.
  }
}

/**
 * Paints a theme without remembering it.
 *
 * What the editor shows while somebody is dragging a colour picker. The
 * difference from [`applyTheme`] is the cache, and it is the whole reason this
 * exists separately: a draft that was cached would be replayed on the next
 * cold start as though it had been saved, so cancelling an edit would still
 * change what the app looks like tomorrow.
 */
export function previewTheme(mode: ResolvedTheme, custom: CustomTheme | null): void {
  document.documentElement.dataset.theme = mode;
  paint(tokensFor(custom, mode));
}

/**
 * The file name a theme called `name` is stored under.
 *
 * A copy of `slug` in `crates/taurus-host/src/theme.rs`, which is the
 * authority — Rust refuses to write a file under any id that does not survive
 * it, so a frontend that guessed differently would show one path in the editor
 * and write to another. `theme.test.ts` checks the two agree, character for
 * character, on the cases that distinguish them.
 */
export function slug(name: string): string {
  let out = "";
  for (const c of name.trim().toLowerCase()) {
    if (/[a-z0-9]/.test(c)) out += c;
    else if (!out.endsWith("-")) out += "-";
  }
  return out.replace(/^-+|-+$/g, "") || "theme";
}

/** The value a token resolves to right now, whoever set it. */
export function currentToken(token: string): string {
  return getComputedStyle(document.documentElement).getPropertyValue(token).trim();
}

/**
 * The palette as it is actually painting — the theme's own values over the
 * stylesheet's, read back off the document.
 *
 * What the contrast check needs, and what "duplicate this into a new theme"
 * starts from. Reading it off the document rather than restating the shipped
 * palette in TypeScript is what keeps this from becoming a third copy of it
 * that drifts: `styles.css` stays the only place those fourteen values live.
 */
export function livePalette(): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [name, token] of Object.entries(COLOR_TOKENS)) out[name] = currentToken(token);
  return out;
}

/** What was showing last time, for the frame before settings arrive. */
export function cachedTheme(): Theme {
  try {
    const stored = localStorage.getItem(CACHE_KEY);
    if (stored === "light" || stored === "dark" || stored === "system") {
      return stored;
    }
  } catch {
    // Same as above: fall through to the system's answer.
  }
  return "system";
}

/**
 * Repaints when the OS flips, while the preference is to follow it.
 *
 * Returns the unsubscribe, so a caller that switches to an explicit theme stops
 * listening rather than leaving a listener that would repaint over the choice.
 *
 * The custom theme has to come back through here with it. An earlier version
 * took only the preference and repainted with `applyTheme("system")`, which
 * cleared every token a custom theme had set — so a branded window reverted to
 * the shipped palette at dusk, and stayed there until something else wrote
 * settings.
 */
export function watchSystemTheme(theme: Theme, custom: CustomTheme | null = null): () => void {
  // Nothing to follow: either the preference is explicit, or the theme paints
  // one mode and the system's opinion cannot change the answer.
  if (theme !== "system" || !window.matchMedia) return () => {};
  if (custom && custom.modes !== "both") return () => {};
  const query = window.matchMedia(DARK_QUERY);
  const onChange = () => applyTheme("system", custom);
  query.addEventListener("change", onChange);
  return () => query.removeEventListener("change", onChange);
}

File: crates/taurus-host/examples/theme.rs
//! What Taurus makes of the themes actually on your disk.
//!
//! ```sh
//! cargo run -p taurus-host --example theme            # global themes
//! cargo run -p taurus-host --example theme -- .       # and this workspace's
//! ```
//!
//! It reads `~/.taurus/themes` and, with a workspace, that folder's
//! `.taurus/themes` as well. It writes nothing.
//!
//! This is the answer to "why is my theme not in the picker", which is the one
//! question this feature will be asked most and the one the app is worst
//! placed to answer — a file that will not parse is, by construction, a file
//! that has nothing to say for itself on screen beyond a line in Settings.
//! Here it says the whole thing: which files were found, which layer each came
//! from, which of the two palettes it can paint, which `--lk-*` tokens it will
//! actually set, whether its logo was read, and every complaint the loader had
//! on the way past.
//!
//! The last of those is the part with no equivalent anywhere else. The app
//! reports the problems of the theme *in force*; the picker reports the rest
//! only while it is open. From a terminal you get all of them at once, before
//! the app has been started.

use std::path::PathBuf;

use taurus_host::theme::{self, COLORS, FONTS};

fn main() {
    let workspace = std::env::args().nth(1).map(PathBuf::from);
    let where_from = workspace
        .as_ref()
        .map(|w| format!("~/.taurus/themes and {}/.taurus/themes", w.display()))
        .unwrap_or_else(|| "~/.taurus/themes".into());
    println!("reading {where_from}\n");

    let (themes, problems) = theme::load_themes(workspace.as_deref());

    if themes.is_empty() {
        println!("No themes. The app paints its own palette, which is not a fault.");
        println!("Writing one: docs/configuration.md#themes\n");
    }

    for theme in &themes {
        let modes = match theme.modes {
            theme::ThemeModes::Both => "dark and light",
            theme::ThemeModes::DarkOnly => "dark only — selecting it pins the mode",
            theme::ThemeModes::LightOnly => "light only — selecting it pins the mode",
        };
        println!("{} ({})", theme.name, theme.id);
        println!("  from    {:?} · {}", theme.scope, theme.path);
        println!("  paints  {modes}");

        for (label, palette) in [("dark", &theme.dark), ("light", &theme.light)] {
            if palette.is_empty() {
                continue;
            }
            println!("  {label}");
            // The token each colour actually sets. A theme names the *job* and
            // the stylesheet names the colour, so this mapping is the one part
            // of a theme file that cannot be checked by reading it.
            for (name, token) in COLORS {
                if let Some(value) = palette.get(*name) {
                    println!("    {name:<14} {value:<10} -> {token}");
                }
            }
        }

        let fonts = [
            (theme.fonts.display.as_deref(), FONTS[0].1),
            (theme.fonts.body.as_deref(), FONTS[1].1),
            (theme.fonts.mono.as_deref(), FONTS[2].1),
        ];
        for (family, token) in fonts {
            if let Some(family) = family {
                // Named rather than checked: whether it is installed is a
                // question only the window can answer, and one it answers
                // silently by falling back.
                println!("  font    {family} -> {token} (must be installed on this machine)");
            }
        }

        match (&theme.wordmark, &theme.logo) {
            (None, None) => {}
            (word, logo) => {
                println!(
                    "  brand   wordmark {} · logo {}",
                    match word.as_deref() {
                        None => "unchanged".to_string(),
                        Some("") => "none, mark only".to_string(),
                        Some(w) => format!("{w:?}"),
                    },
                    match logo {
                        // The inlined length, because that is what travels
                        // with every status the window is pushed. In bytes
                        // under a kilobyte: a wordmark SVG is often 300 bytes,
                        // and "0KB" reads like it failed.
                        Some(uri) if uri.len() < 1024 => {
                            format!("read, {} bytes inlined", uri.len())
                        }
                        Some(uri) => format!("read, {}KB inlined", uri.len() / 1024),
                        None => "none".to_string(),
                    }
                );
            }
        }

        let shape = &theme.shape;
        if shape.radius.is_some() || shape.gutter.is_some() || shape.rail_gutter.is_some() {
            println!(
                "  shape   radius {} · gutter {} · rail-gutter {}",
                shape
                    .radius
                    .map_or("as shipped".into(), |r| format!("{r}x")),
                shape
                    .gutter
                    .map_or("as shipped".into(), |g| format!("{g}px")),
                shape
                    .rail_gutter
                    .map_or("as shipped".into(), |g| format!("{g}px")),
            );
        }
        println!();
    }

    if problems.is_empty() {
        println!("Nothing to report.");
    } else {
        println!("{} problem{}:", problems.len(), plural(problems.len()));
        for problem in &problems {
            println!("  {problem}");
        }
        println!(
            "\nA file that will not parse costs itself and nothing else — everything above still loads."
        );
    }

    // Contrast is deliberately not scored here. It is checked in the editor,
    // against the palette *resolved* over whichever of the two the app ships,
    // and the stylesheet those defaults live in is not something Rust can
    // read. A second table of pairs here would be a second thing to keep in
    // step with `src/lib/contrast.ts`, and the one that drifted would be this
    // one — because nothing looks at it as often.
    println!("\nContrast is checked in Settings › Appearance, where the palette can be");
    println!("resolved against the one the app ships.");
}

fn plural(n: usize) -> &'static str {
    if n == 1 {
        ""
    } else {
        "s"
    }
}
